<HTML>
<HEAD>
<meta http-equiv="Content-Type" content="text/html; charset=us-ascii">
<TITLE>
    CWG Issue 668</TITLE>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<STYLE TYPE="text/css">
  INS { text-decoration:none; font-weight:bold; background-color:#A0FFA0 }
  .INS { text-decoration:none; background-color:#D0FFD0 }
  DEL { text-decoration:line-through; background-color:#FFA0A0 }
  .DEL { text-decoration:line-through; background-color: #FFD0D0 }
  @media (prefers-color-scheme: dark) {
    HTML { background-color:#202020; color:#f0f0f0; }
    A { color:#5bc0ff; }
    A:visited { color:#c6a8ff; }
    A:hover, a:focus { color:#afd7ff; }
    INS { background-color:#033a16; color:#aff5b4; }
    .INS { background-color: #033a16; }
    DEL { background-color:#67060c; color:#ffdcd7; }
    .DEL { background-color:#67060c; }
  }
  SPAN.cmnt { font-family:Times; font-style:italic }
</STYLE>
</HEAD>
<BODY>
<P><EM>This is an unofficial snapshot of the ISO/IEC JTC1 SC22 WG21
  Core Issues List revision 116c.
  See http://www.open-std.org/jtc1/sc22/wg21/ for the official
  list.</EM></P>
<P>2025-03-08</P>
<HR>
<A NAME="668"></A><H4>668.
  
Throwing an exception from the destructor of a local static object
</H4>
<B>Section: </B>14.6.2&#160; [<A href="https://wg21.link/except.terminate">except.terminate</A>]
 &#160;&#160;&#160;

 <B>Status: </B>CD2
 &#160;&#160;&#160;

 <B>Submitter: </B>Daniel Kr&#252;gler
 &#160;&#160;&#160;

 <B>Date: </B>16 December 2007<BR>


<P>[Voted into the WP at the March, 2009 meeting.]</P>

<P>The destruction of local static objects occurs at the same time as
that of non-local objects (6.9.3.3 [<A href="https://wg21.link/basic.start.dynamic#1">basic.start.dynamic</A>] paragraph 1) and
the execution of functions registered with <TT>std::atexit</TT>
(paragraph 3).  According to 14.6.2 [<A href="https://wg21.link/except.terminate#1">except.terminate</A>] paragraph 1,
<TT>std::terminate</TT> is called if a destructor for a non-local
object or a function registered with <TT>std::atexit</TT> exits via
an exception, but the Standard is silent about the result of throwing
an exception from a destructor for a local static object.  Presumably
this is an oversight and the same rules should apply to destruction of
local static objects.</P>

<P><B>Proposed resolution (September, 2008):</B></P>

<P>Change 14.6.2 [<A href="https://wg21.link/except.terminate#1">except.terminate</A>] paragraph 1, fourth bullet as
indicated, and add an additional bullet to follow it:</P>

<UL>
<LI><P>when construction <DEL>or destruction</DEL> of a non-local
object with static or thread storage duration exits using an
exception (6.9.3.2 [<A href="https://wg21.link/basic.start.static">basic.start.static</A>]), or</P></LI>

<LI><P><INS>when destruction of an object with static or thread
storage duration exits using an exception (6.9.3.3 [<A href="https://wg21.link/basic.start.dynamic">basic.start.dynamic</A>]), or</INS></P></LI> </UL>

<BR><BR>
</BODY>
</HTML>
